### PR TITLE
SOLR-16507: Refactor out ImplicitSnitch

### PR DESCRIPTION
**Pull request by noblepaul**
This is work in progress

### CODE REVIEW

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/MetricsFetcher.java
Comment (dsmiley) [inline solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/MetricsFetcher.java:22]: Javadocs please.  If it was written for something in particular in mind, then it's very useful to say that, even if it may theoretically have future uses we don't know about.  It's not clear why would something use this instead of going directly to the Metrics API?  As a newcomer to this code, is it related to having short/simply well-known named "tags" (formerly maybe called snitch?) that loosely correlate to a more complex metrics expression, so that some other code can use these tags to make node placement decisions based on those same tags?

The lifecycle of an instance of this class is not obvious; please add a sentence on that.

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/MetricsFetcher.java
Comment (dsmiley) [inline solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/MetricsFetcher.java:177]: At least a one line javadoc on what this is.  Seems similar to the other "tags" although the value varies?  Say that.

**Review by dsmiley — CHANGES_REQUESTED** [review]
Looks like a vast improvement; thanks!

Is an objective here to remove the word "snitch"?  Unless you think it holds value, I'd be in favor of that.  I see it still used though.

**Comment by noblepaul** [comment]
> Looks like a vast improvement; thanks!
> 
> Is an objective here to remove the word "snitch"? Unless you think it holds value, I'd be in favor of that. I see it still used though.

That too and get rid of a lot of classes because we just need a mechanism to fetch Per node details. Maybe we need to find a better term , but "snitch" must go

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/SolrClientNodeStateProvider.java
Comment (dsmiley) [inline solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/SolrClientNodeStateProvider.java:225]: Can you please add at least one javadoc line explaining what this is for?  I don't get it.  It seems to be the "Command" software pattern?  Maybe call it "RemoteCallCmd" if so.

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/SolrClientNodeStateProvider.java
Comment (dsmiley) [inline solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/SolrClientNodeStateProvider.java:63]: Can you please add a comment illustrating why the keys & values look like?

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java
Comment (dsmiley) [inline solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java:41]: Still looking for a lifecycle like bit of documentation.   It has no state; right?  In that case, all methods should probably be static and we say this is a "utility" like class and document it's not instantiated.

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java
Comment (dsmiley) [inline solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java:103]: is not subclassed; should be private

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java
Comment (dsmiley) [reply solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java:38]: Can you please respond to my comment @noblepaul ?

Before merging PRs, please wait for either an explicit approval or announce a 2-day plan to merge ("lazy consensus").

Comment (noblepaul) [reply solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/NodeValueFetcher.java:38]: > It has no state; right? In that case, all methods should probably be static 

Can it be made static? Yes, But we could potentially add some state later. That's why I left it as is. However, please feel free to make any change to this. That's why I made this branch is the public repo instead of my own

**Pull request merged**
